python/cudf/cudf/core/dtypes.py

```diff
@@ -319,16 +319,36 @@ def __eq__(self, other: Dtype) -> bool:
         elif other.ordered is None and other._categories is None:
             # other is equivalent to the string "category"
             return True
-        elif self.ordered != other.ordered:
-            return False
         elif self._categories is None or other._categories is None:
-            return True
+            return self._categories is other._categories
+        elif self.ordered or other.ordered:
+            return (self.ordered == other.ordered) and self._categories.equals(
+                other._categories
+            )
         else:
-            return (
-                self._categories.dtype == other._categories.dtype
-                and self._categories.equals(other._categories)
+            left_cats = self._categories
```

Comment:
Why wasn't `self._categories.equals` a sufficient check here? That's what the old code deferred to. Is the issue the need for the strict vs non-strict check for unordered categories? If so, couldn't that be handled by just doing the sort before calling the `equals` method?

Reply:
That wasn't alone the bug in this equality check. The ordering of the `elif ordered condition` too is a bug. In the new `else` block we are also pre-emptively returning `False` if the size's don't match.